src/macroxs_operations.F90

```diff
@@ -0,0 +1,243 @@
+module macroxs_operations
+
+  use constants
+  use macroxs_header,  only: MacroXS, MacroXSIso, MacroXSAngle, &
+                             expand_harmonic
+  use material_header, only: Material
+  use math
+  use nuclide_header,  only: find_angle, MaterialMacroXS, NuclideMicroXS, &
+                             NuclideMG, NuclideMGContainer
+  use random_lcg,      only: prn
+  use scattdata_header
+  use search
+
+  implicit none
+
+contains
+
+!===============================================================================
+! UPDATE_XS stores the xs to work with
+!===============================================================================
+
+  subroutine calculate_mgxs(this, gin, uvw, xs)
+    class(MacroXS),        intent(in)    :: this
+    integer,               intent(in)    :: gin         ! Incoming neutron group
+    real(8),               intent(in)    :: uvw(3)      ! Incoming neutron direction
+    type(MaterialMacroXS), intent(inout) :: xs
+
+    integer :: iazi, ipol
+
+    select type(this)
+    type is (MacroXSIso)
+      xs % total         = this % total(gin)
+      xs % elastic       = this % scattxs(gin)
+      xs % absorption    = this % absorption(gin)
+      xs % nu_fission    = this % nu_fission(gin)
+
+    type is (MacroXSAngle)
+      call find_angle(this % polar, this % azimuthal, uvw, iazi, ipol)
+      xs % total         = this % total(gin, iazi, ipol)
+      xs % elastic       = this % scattxs(gin, iazi, ipol)
+      xs % absorption    = this % absorption(gin, iazi, ipol)
+      xs % nu_fission    = this % nu_fission(gin, iazi, ipol)
+    end select
+
+  end subroutine calculate_mgxs
+
+
+!===============================================================================
+! SAMPLE_FISSION_ENERGY acts as a templating code for macroxs_*_sample_fission_energy
+!===============================================================================
+
+  function sample_fission_energy(this, gin, uvw) result(gout)
+    class(MacroXS),      intent(in) :: this   ! Data to work with
+    integer, intent(in)             :: gin    ! Incoming energy group
+    real(8), intent(in)             :: uvw(3) ! Particle Direction
+    integer                         :: gout   ! Sampled outgoing group
+
+    select type(this)
+    type is (MacroXSIso)
+      gout = macroxsiso_sample_fission_energy(this, gin, uvw)
+    type is (MacroXSAngle)
+      gout = macroxsangle_sample_fission_energy(this, gin, uvw)
+    end select
```

Comment:
You could make sample_fission_energy a deferred type-bound procedure of MacroXS instead of having this dispatch code here. Was there a reason you wanted to use select type as opposed to runtime polymorphism?